Adds new specutils data loader for SDSS MaNGA cube/rss data

havok2063 opened the pull request:
This PR adds a new data loader for SDSS MaNGA cube and rss spectral data.  It also adds a convenience function, `io.registers.identify_spectrum_format` to identify a valid registered `specutils` file format given an input filename.  This function is essentially a wrapper for `astropy.io.registry.identify_format` setting origin to `read` and data_class to `Spectrum1D`.

nmearl reviewed (dismissed):
Thanks, @havok2063, this looks great. Couple of small changes, but also one bigger change to suggest as well.

We've moved to allowing either string file paths or file object references in data loading. E.g. if you look at the [hst loaders](https://github.com/astropy/specutils/blob/master/specutils/io/default_loaders/hst_cos.py#L38), you'll see that they handle the case where a file object is passed to the reader. It'd be great if you could update these to handle that case as well.

havok2063 commented:
Thanks for the comments @nmearl.  I've accepted your minor suggestions.  I've updated the manga loaders to use the `file_obj` nomenclature following the hst examples.  I originally followed the jwst example, which look like they haven't been updated either.

dhomeier commented:
Thanks for the update! Ideally accepting file handles or URLs would also work for the identifier, using information from the file content itself – that is essentially already implemented here, but will require slightly different treatment depending on the type of `file_obj`.
I suggest to have a look at e.g. `spec_identify` in the SDSS loader; you probably just need to add the check for `header['INSTRUME'] == 'MaNGA'` there.
Since the `identify_spectrum_format` function is not directly used in the loader, could you add some documentation when and how to use it?

The test failure seems due to an unrelated download failure in testing another loader, but it would be great if you could also add a test for the new format(s), assuming there are some example files on sdss.org (see e.g. `test_sdss_spec`).

havok2063 commented:
@dhomeier Thanks for the comments.  I've updated the code so the identifier functions also accept a file object.  I've added some documentation on the `identify_spectrum_format` function.  I created a new page, `identify.html`, but am happy to incorporate it into a better spot.  I've also added two new tests `test_manga_cube` and `test_manga_rss`.  These simply pass in the public url to the data into `Spectrum1D.read`.  I tried following the old sdss test with `urllib` but ran into issues where Spectrum1D could not read the objects from the `urllib` response.   I was getting strange errors regarding `header does not have valid END card`.  I tried setting `ignore_missing_end=True` but then got an error about `empty or corrupt FITS file`.  When I download the data files locally or use the url stream directly, `Spectrum1D.` reads the file just fine.

dhomeier commented on docs/identify.rst:
```diff
@@ -0,0 +1,40 @@
+    >>> from specutils.io.registers import identify_spectrum_format
+    >>>
+    >>> url='https://dr15.sdss.org/sas/dr15/manga/spectro/redux/v2_4_3/8485/stack/manga-8485-1901-LOGCUBE.fits.gz'
+    >>> dd = download_file(url)
```
```suggestion
    >>> dd = download_file(url)  # doctest: +SKIP
```
This made the doctest fail when not run with `remote_data=True` – don't know if one could add the `remote_data` decorator somehow for doctest, but I'd suggest to simply skip it.

dhomeier commented on docs/identify.rst:
```diff
@@ -0,0 +1,40 @@
+
+This function is useful for identifying a spectrum file format without the need to
+read it in with Spectrum1d's `~specutils.Spectrum1D.read` method.  It returns the best guess as to a
+valid format from the list of ``Formats`` as given by `~astropy.io.registry.get_formats`.
```
Perhaps this could spell out still more clearly that it is using the same identification methods as `Spectrum1D.read` – I doubt many users will be familiar with the internals of `~astropy.io.registry`.

havok2063 replied:
I've updated the docs here with another sentence or two.  Let me know what you think.

dhomeier commented on specutils/io/default_loaders/manga.py:
```diff
@@ -0,0 +1,177 @@
+    """
+    is_sdss = _identify_sdss_fits(*args, **kwargs)
+    with _read_fileobj(*args, **kwargs) as hdulist:
+        return (is_sdss and "FLUX" in hdulist and hdulist[1].header['INSTRUME'] == 'MaNGA'
```
Wondering if it's really worth delegating the `is_sdss` check to its own function, as the direct `hdulist[0].header["TELESCOP"] == "SDSS 2.5-M"` test would fit almost within the same space.
Since it is obviously working, it should be OK, but we might want to avoid extra file I/O operations from a performance point of view.

havok2063 replied:
I agree.  I've removed the redundant function and moved the the SDSS telescope check into the `cube` and `rss` identifiers

dhomeier commented on specutils/io/default_loaders/manga.py:
```diff
@@ -0,0 +1,177 @@
+
+    Parameters
+    ----------
+    filename : str
+        The path to the FITS file
```
```suggestion
    file_obj: str, file-like, or HDUList
          FITS file name, object (provided from name by Astropy I/O Registry),
          or HDUList (as resulting from astropy.io.fits.open()).
```

dhomeier commented on specutils/io/default_loaders/manga.py:
```diff
@@ -0,0 +1,177 @@
+
+    Parameters
+    ----------
+    filename : str
+        The path to the FITS file
```
```suggestion
    file_obj: str, file-like, or HDUList
          FITS file name, object (provided from name by Astropy I/O Registry),
          or HDUList (as resulting from astropy.io.fits.open()).
```

dhomeier commented on specutils/io/registers.py:
```diff
@@ -167,3 +169,37 @@ def _load_user_io():
+    Given a filename, attempts to identify a valid file format
+    from the list of registered specutils loaders.  Essentially a wrapper for
+    `~astropy.io.registry.identify_format` setting **origin** to ``read`` and
+    **data_class_required** to `~specutils.Spectrum1D`.
```
I am curious if this can be generalised to other `spectra` objects, e.g. by adding a `dtype=Spectrum1D` kwarg.
There are one or two loaders for `SpectrumList` now, and #708 should add one for `SpectrumCollection`, so it would be great if the same function could be used for identification. Does not have to be solved in this PR, but then the documentation would also need to be updated a little bit.

havok2063 replied:
I made a small change to allow a user to specify a dtype that can either be `Spectrum1D`, `SpectrumList`, or `SpectrumCollection`, with the default being `Spectrum1D`.  It throws an assert otherwise.

dhomeier commented:
> incorporate it into a better spot. I've also added two new tests `test_manga_cube` and `test_manga_rss`. These simply pass in the public url to the data into `Spectrum1D.read`. I tried following the old sdss test with `urllib` but ran into issues where Spectrum1D could not read the objects from the `urllib` response. I was getting strange errors regarding `header does not `

Thanks a lot for the updates and sorry for pointing to a misleading example; I should have referred to the `test_*_stream` tests right away!

I especially like the implementation with the custom contextmanager – I really think this could be used by all loaders supporting all 3 `file_obj` types, so perhaps `_read_fileobj` could be moved out to `parsing_utils.py`. What do you think @nmearl ?

nmearl commented:
> I especially like the implementation with the custom contextmanager – I really think this could be used by all loaders supporting all 3 file_obj types, so perhaps _read_fileobj could be moved out to parsing_utils.py (maybe renamed to something like read_fileobj_or_hdulist). What do you think @nmearl ?

Yes, definitely. It's probably not necessary to do that here in this PR, but we should certainly open an issue to update the docs and the identifier/loading functions to use a generic context manager.

havok2063 commented:
@dhomeier Thanks for the comments.  I've pushed the changes you've requested.

eteq commented on docs/identify.rst:
```diff
@@ -0,0 +1,43 @@
+`~specutils.Spectrum1D` file format from the list of registered formats, and
+essentially acts as a wrapper on `~astropy.io.registry.identify_format`.
+
+This function is useful for identifying a spectrum file format without the need to
+read it in with Spectrum1d's `~specutils.Spectrum1D.read` method.  It uses the
```
```suggestion
This function is useful for identifying a spectrum file format without reading the
whole file with the  `~specutils.Spectrum1D.read` method.  It uses the
```

eteq commented on docs/identify.rst:
```diff
@@ -0,0 +1,43 @@
+
+This function is useful for identifying a spectrum file format without the need to
+read it in with Spectrum1d's `~specutils.Spectrum1D.read` method.  It uses the
+same identification method as ``read`` however, so it simply provides a convenience
```
```suggestion
same identification method as ``read`` however, so it provides a convenience
```

eteq commented on docs/identify.rst:
```diff
@@ -0,0 +1,43 @@
+It returns the best guess as to a valid format from the list of ``Formats``
+as given by `~astropy.io.registry.get_formats`.
+
+General usage is as follows, passing in an absolute filename path:
+
+.. code-block:: python
+
+    >>> from specutils.io.registers import identify_spectrum_format
+    >>> identify_spectrum_format("/path/to/file.fits")  # doctest: +SKIP
+
+An example is given to identify a SDSS MaNGA data cube file:
+
```
I think the "general usage" isn't needed given that the MaNGA example is already quite clear
```suggestion
For eample, to identify a SDSS MaNGA data cube file:

```

eteq commented on specutils/io/registers.py:
```diff
@@ -167,3 +169,45 @@ def _load_user_io():
+    Parameters
+    ----------
+    filename : str
+        The absolute filename to the object
```
it doesn't *have* to be absolute, right?  i.e. `../some-file.fits` should work even if it's not the best idea to do that?
```suggestion
        A path to a file to be identified
```

eteq commented on specutils/io/default_loaders/manga.py:
```diff
@@ -0,0 +1,141 @@
+    else:
+        hdulist = fits.open(file_obj, **kwargs)
+
+    spaxel = u.Unit('spaxel', represents=u.pixel, doc='0.5" spatial pixel', parse_strict='silent')
```
This is actually an interesting and good use of units.  I wonder if this should be moved to the public API so that users can do e.g. `specutils.<something>.spaxel`? 

(this could be done in a folllow-on PR if you don't want to deal with it right now though - I can make an issue to deal with later if that's your preference @havok2063)

eteq commented on specutils/io/default_loaders/manga.py:
```diff
@@ -0,0 +1,141 @@
+    else:
+        hdulist = fits.open(file_obj, **kwargs)
+
+    fiber = u.Unit('fiber', represents=u.pixel, doc='spectroscopic fiber', parse_strict='silent')
```
(ditto for the spaxel comment above)

eteq reviewed (approved):
This looks good as it stands, although I have several minor comments that I'm am suggesting to @havok2063 that he can take or leave depending on preference.

One higher-level thought (which doesn't need to hold up this PR, but might be easy to include) - should the rss format *also* have a `SpectrumCollection` option? I'm not familiar enough with it to know for sure, but it looks like it might be a case where it's more useful to think of it as a set of distinct fibers instead of spaxels?

nmearl commented:
> One higher-level thought (which doesn't need to hold up this PR, but might be easy to include) - should the rss format also have a SpectrumCollection option?

Just want to point out an issue here that `SpectrumCollection` isn't not a subclass of `NDData`, and thus doesn't leverage the astropy io machinery. We'd have to refactor `SpectrumCollection` a bit to allow e.g. `SpectrumCollection.read(...)`.

However, we _can_ still do `Spectrum1D.read(...)` and have it return a `SpectrumCollection` object, but it might be too inconsistent of a user workflow.

havok2063 commented:
@eteq  Well admittedly I've always been a bit confused between some of the spectrum class distinctions outlined in the docs.  I was under the impression that `SpectrumCollection` was for a set of independent spectra that have different spectral axes, either a different length or different WCSes, and the FITS contains individual spectral axes for each spectrum.    If that's not the case, then yeah technically rss fibers could be a `SpectrumCollection`.  MaNGA rss files contain a set of fiber spectra (in a 2d-array) that all have the same spectral axis, i.e. same length and WCS.  There's one FITS extension containing the wavelength array for all spectral fibers.  The fibers aren't exactly independent of each other.  For example, if you plotted them on sky, you'd get a series of overlapping dithered fiber positions centered on a target.   It's the same situation as the Cubes, which I also read in as `Spectrum1D`.   I could be convinced of changing it, but less so if it requires some refactoring.  To me when I hear "SpectrumCollection" I think "a loose collection of random spectra I want to treat together".   I'd be happy to make this change in a future PR.

I've accepted your small changes.  They make sense to me.   Regarding the units, what do you mean by moving to the public API?  Does that mean moving these definitions into `astropy.units` or is this something that can be done within `specutils`?  If it's the latter, I can probably easily do that now.  If it's the former I might prefer creating a separate issue and punt this to a later PR since that means going through a whole new process.

dhomeier commented:
> Just want to point out here that `SpectrumCollection` isn't a subclass of `NDData`, and thus doesn't leverage the astropy io machinery. We'd have to refactor `SpectrumCollection` a bit to allow e.g. `SpectrumCollection.read(...)`.

That is just what I have done in #702 (by simply subclassing `SpectrumCollection` as`NDIOMixin`).

> @eteq Well admittedly I've always been a bit confused between some of the spectrum class distinctions outlined in the docs. I was under the impression that SpectrumCollection was for a set of independent spectra that have different spectral axes, either a different length or different WCSes, and the FITS contains individual spectral axes for each spectrum. If that's not the case, then yeah technically rss fibers could be a SpectrumCollection.

`SpectrumCollection` still is a relatively structured object, notably requiring all spectral axes to have the same length, but with generally different values. The loosest collection really is the `SpectrumList`.
But if your spectra have indeed all _identical_ `spectral_axis`, I agree that `Spectrum1D` would be the natural representation.

havok2063 commented:
ahh ok, that makes sense.  I hadn't realized that subtle distinction between `SpectrumCollection` and SpectrumList`.  Thanks for clarifying.  Yeah the MaNGA data all have identical spectral axes in length and values.  It sounds like `Spectrum1D` is the right thing for them to be but happy to hear arguments otherwise.  If everyone agrees, then I think this is ready to be merged, and @eteq can create a new issue to flesh out what should be done with the units.

eteq commented:
Alright, yep I think all is good, and I'll make a separate issue for the units thing
